File: sandbox/python/sandbox.py
from dataclasses import dataclass
from typing import Any, Callable, Mapping


@dataclass(frozen=True)
class ToolRequest:
    name: str
    arguments: Mapping[str, Any]


class SandboxError(ValueError):
    pass


class AiSandbox:
    """Policy layer, not OS isolation. Keep callbacks outside the trust boundary."""

    def __init__(self, allowed_tools: set[str], *, max_steps: int = 8, max_argument_bytes: int = 16_384):
        if max_steps < 1 or max_argument_bytes < 1:
            raise ValueError("limits must be positive")
        self.allowed_tools = frozenset(allowed_tools)
        self.max_steps = max_steps
        self.max_argument_bytes = max_argument_bytes
        self.steps = 0

    def run(self, request: ToolRequest, executor: Callable[[ToolRequest], Any]) -> Any:
        if request.name not in self.allowed_tools:
            raise SandboxError("tool is not allowed")
        if self.steps >= self.max_steps:
            raise SandboxError("step budget exceeded")
        if len(repr(dict(request.arguments)).encode()) > self.max_argument_bytes:
            raise SandboxError("argument budget exceeded")
        self.steps += 1
        return executor(request)

File: sandbox/javascript/sandbox.mjs
export class SandboxError extends Error {}

export class AiSandbox {
  constructor(allowedTools, { maxSteps = 8, maxArgumentBytes = 16_384 } = {}) {
    if (maxSteps < 1 || maxArgumentBytes < 1) throw new RangeError("limits must be positive");
    this.allowedTools = new Set(allowedTools);
    this.maxSteps = maxSteps;
    this.maxArgumentBytes = maxArgumentBytes;
    this.steps = 0;
  }

  authorize(request) {
    if (!this.allowedTools.has(request.name)) throw new SandboxError("tool is not allowed");
    if (this.steps >= this.maxSteps) throw new SandboxError("step budget exceeded");
    const encoded = JSON.stringify(request.arguments ?? {});
    if (Buffer.byteLength(encoded, "utf8") > this.maxArgumentBytes) throw new SandboxError("argument budget exceeded");
    this.steps += 1;
    return Object.freeze({ name: request.name, arguments: request.arguments ?? {} });
  }
}
